k8ssandra-operator in CrashLoopBackOff during Kubernetes upgrade

**What happened?**

We had a K8ssandra cluster with Medusa enabled deployed in AKS 1.29.
Cluster admin upgraded AKS to 1.30, and after that k8ssandra-operator was in `CrashLoopBackOff` and the K8ssandra pods were not starting.
K8ssandra pods started finally 4h30 after the AKS upgrade.


**Did you expect to see something different?**

I expect a Kubernetes upgrade to not causes k8ssandra unavailibity.

**How to reproduce it (as minimally and precisely as possible)**:

1. Deploy an AKS cluster 1.29
2. Deploy k8ssandra-operator v1.14
3. Deploy a K8ssandra cluster with at least 3 replicas and Medusa backup enabled
4. Create a MedusaBackupSchedule or create MedusaBackupJobs
5. Delete the nodepool containing Cassandra pods, so that all Cassandra pods will be evicted
6. k8ssandra-operator should crash when it attempts to reconcile a `MedusaBackupJob` for this k8ssandra cluster
7. Create a new 1.30 nodepool

**Environment**

* K8ssandra Operator version:
1.14.0

* Kubernetes version information:
1.29 -> 1.30

* Kubernetes cluster kind:
AKS

* Manifests:

K8ssandraCluster manifest (after recovery):
```
apiVersion: k8ssandra.io/v1alpha1
kind: K8ssandraCluster
metadata:
  annotations:
    k8ssandra.io/initial-system-replication: '{"cs-1855ea8cc2":3}'
  creationTimestamp: "2023-08-15T18:06:51Z"
  finalizers:
  - k8ssandracluster.k8ssandra.io/finalizer
  generation: 3236
  name: cs-1855ea8cc2
  namespace: qaazv8dow
  resourceVersion: "348869397"
  uid: 6ffc08c2-d358-4f3b-a102-e99dbd396651
spec:
  auth: true
  cassandra:
    datacenters:
    - config:
        cassandraYaml:
          audit_logging_options:
            enabled: true
          commitlog_directory: /c3/cassandra/data/commitlog
          data_file_directories:
          - /c3/cassandra/data/data
          saved_caches_directory: /c3/cassandra/data/saved_caches
        jvmOptions:
          gc: G1GC
          heapSize: 12G
      containers:
      - args:
        - -c
        - tail -n+1 -F /c3/cassandra/logs/system.log
        command:
        - /bin/sh
        image: prodqaacr.azurecr.io/bundled-cassandra-jre-bcfips:4.0.12-r1-202407171520
        imagePullPolicy: Always
        name: server-system-logger
        resources: {}
        volumeMounts:
        - mountPath: /c3/cassandra
          name: server-data
      - args:
        - mgmtapi
        command:
        - /docker-entrypoint.sh
        name: cassandra
        resources: {}
        volumeMounts:
        - mountPath: /c3/cassandra
          name: server-data
      - env:
        - name: MEDUSA_TMP_DIR
          value: /c3/cassandra/medusa/tmp
        name: medusa
        resources: {}
        volumeMounts:
        - mountPath: /c3/cassandra
          name: server-data
      initContainers:
      - command:
        - /usr/local/bin/entrypoint
        image: prodqaacr.azurecr.io/bundled-cassandra-jre-bcfips:4.0.12-r1-202407171520
        imagePullPolicy: Always
        name: server-config-init
        resources: {}
      - env:
        - name: MEDUSA_TMP_DIR
          value: /c3/cassandra/medusa/tmp
        name: medusa-restore
        resources: {}
        volumeMounts:
        - mountPath: /c3/cassandra
          name: server-data
      metadata:
        name: cs-1855ea8cc2
        pods:
          labels:
            azure.workload.identity/use: "true"
            c3__app-0: 0c30
            c3__app_id-0: 0qaazv8dow-c3-c30
            c3__cluster-0: 0qaazv8dow0
            c3__created-0: 02023-12-13T00_3A42_3A42.127Z0
            c3__created_by-0: 0worker0
            c3__created_from-0: 0qaazv8dow-c3-c3-k8spo-appleader-001-6648dbd5cbsq4hl0
            c3__env-0: 0c30
            c3__env_id-0: 0qaazv8dow-c30
            c3__func-0: 0k8sc3cass0
            c3__id-0: 0qaazv8dow-c3-c3-k8sc3cass0
            c3__namespace-0: 0qaazv8dow0
            c3__role-0: "00"
            c3__seq-0: "00"
            c3__service-0: 0qaazv8dow-c3-c3-k8scass-cs-0010
            c3__subseq-0: "00"
            c3__updated-0: 02023-12-13T00_3A42_3A42.127Z0
            c3__updated_by-0: 0worker0
            c3__updated_from-0: 0qaazv8dow-c3-c3-k8spo-appleader-001-6648dbd5cbsq4hl0
        services:
          additionalSeedService: {}
          allPodsService: {}
          dcService: {}
          nodePortService: {}
          seedService: {}
      perNodeConfigInitContainerImage: mikefarah/yq:4
      perNodeConfigMapRef: {}
      podSecurityContext:
        fsGroup: 999
        runAsGroup: 999
        runAsUser: 999
      serviceAccount: c3cassandra
      size: 3
      stargate:
        allowStargateOnDataNodes: false
        containerImage:
          name: stargate-v1
          registry: prodqaacr.azurecr.io
          repository: c3.ai
          tag: 1.0.85-r2-202410132043
        heapSize: 256M
        metadata:
          commonLabels:
            azure.workload.identity/use: "true"
            c3__app-0: 0c30
            c3__app_id-0: 0qaazv8dow-c3-c30
            c3__cluster-0: 0qaazv8dow0
            c3__created-0: 02023-12-13T00_3A42_3A42.127Z0
            c3__created_by-0: 0worker0
            c3__created_from-0: 0qaazv8dow-c3-c3-k8spo-appleader-001-6648dbd5cbsq4hl0
            c3__env-0: 0c30
            c3__env_id-0: 0qaazv8dow-c30
            c3__func-0: 0k8sc3cass0
            c3__id-0: 0qaazv8dow-c3-c3-k8sc3cass0
            c3__namespace-0: 0qaazv8dow0
            c3__role-0: "00"
            c3__seq-0: "00"
            c3__service-0: 0qaazv8dow-c3-c3-k8scass-cs-0010
            c3__subseq-0: "00"
            c3__updated-0: 02023-12-13T00_3A42_3A42.127Z0
            c3__updated_by-0: 0worker0
            c3__updated_from-0: 0qaazv8dow-c3-c3-k8spo-appleader-001-6648dbd5cbsq4hl0
          pods: {}
          service: {}
        secretsProvider: internal
        serviceAccount: c3cassandra
        size: 1
        tolerations:
        - effect: NoSchedule
          key: cassandra
          operator: Equal
          value: "true"
      stopped: false
      storageConfig:
        cassandraDataVolumeClaimSpec:
          accessModes:
          - ReadWriteOnce
          resources:
            requests:
              storage: 1Ti
          storageClassName: default
      tolerations:
      - effect: NoSchedule
        key: cassandra
        operator: Equal
        value: "true"
    jmxInitContainerImage:
      name: busybox
      registry: docker.io
      tag: 1.34.1
    metadata:
      pods: {}
      services:
        additionalSeedService: {}
        allPodsService: {}
        dcService: {}
        nodePortService: {}
        seedService: {}
    perNodeConfigInitContainerImage: mikefarah/yq:4
    resources:
      limits:
        cpu: "8"
        memory: 64Gi
      requests:
        cpu: "8"
        memory: 64Gi
    serverType: cassandra
    serverVersion: 4.0.13
    superuserSecretRef: {}
    telemetry:
      mcac:
        enabled: true
        metricFilters:
        - deny:org.apache.cassandra.metrics.Table
        - deny:org.apache.cassandra.metrics.table
        - allow:org.apache.cassandra.metrics.table.live_ss_table_count
        - allow:org.apache.cassandra.metrics.Table.LiveSSTableCount
        - allow:org.apache.cassandra.metrics.table.live_disk_space_used
        - allow:org.apache.cassandra.metrics.table.LiveDiskSpaceUsed
        - allow:org.apache.cassandra.metrics.Table.Pending
        - allow:org.apache.cassandra.metrics.Table.Memtable
        - allow:org.apache.cassandra.metrics.Table.Compaction
        - allow:org.apache.cassandra.metrics.table.read
        - allow:org.apache.cassandra.metrics.table.write
        - allow:org.apache.cassandra.metrics.table.range
        - allow:org.apache.cassandra.metrics.table.coordinator
        - allow:org.apache.cassandra.metrics.table.dropped_mutations
        - allow:org.apache.cassandra.metrics.Table.TombstoneScannedHistogram
        - allow:org.apache.cassandra.metrics.table.tombstone_scanned_histogram
      prometheus:
        enabled: true
  medusa:
    cassandraUserSecretRef: {}
    certificatesSecretRef: {}
    containerImage:
      name: c3-cassandra-medusa-fips
      pullPolicy: Always
      registry: prodqaacr.azurecr.io
      repository: c3.ai
      tag: 0.21.0-r5-202407181745
    medusaConfigurationRef: {}
    storageProperties:
      bucketName: qaazv8dow
      concurrentTransfers: 0
      maxBackupAge: 7
      maxBackupCount: 0
      multiPartUploadThreshold: 104857600
      prefix: c3cassandra-backup
      storageProvider: azure_blobs
      storageSecretRef:
        name: cs-1855ea8cc2-medusa-bucket-key
      transferMaxBandwidth: 50MB/s
  reaper:
    ServiceAccountName: c3cassandra
    autoScheduling:
      enabled: true
      initialDelayPeriod: PT15S
      percentUnrepairedThreshold: 10
      periodBetweenPolls: PT10M
      repairType: AUTO
      scheduleSpreadPeriod: PT6H
      timeBeforeFirstSchedule: PT5M
    cassandraUserSecretRef:
      name: cs-1855ea8cc2-superuser
    containerImage:
      name: cassandra-reaper-jre-bcfips
      registry: prodqaacr.azurecr.io
      repository: c3.ai
      tag: 3.5.0-202403142052
    deploymentMode: PER_DC
    heapSize: 2Gi
    httpManagement:
      enabled: false
    initContainerImage:
      name: cassandra-reaper-jre-bcfips
      registry: prodqaacr.azurecr.io
      repository: c3.ai
      tag: 3.5.0-202403142052
    jmxUserSecretRef: {}
    keyspace: reaper_db
    metadata:
      pods:
        labels:
          azure.workload.identity/use: "true"
          c3__app-0: 0c30
          c3__app_id-0: 0qaazv8dow-c3-c30
          c3__cluster-0: 0qaazv8dow0
          c3__created-0: 02023-12-13T00_3A42_3A42.127Z0
          c3__created_by-0: 0worker0
          c3__created_from-0: 0qaazv8dow-c3-c3-k8spo-appleader-001-6648dbd5cbsq4hl0
          c3__env-0: 0c30
          c3__env_id-0: 0qaazv8dow-c30
          c3__func-0: 0k8sc3cass0
          c3__id-0: 0qaazv8dow-c3-c3-k8sc3cass0
          c3__namespace-0: 0qaazv8dow0
          c3__role-0: "00"
          c3__seq-0: "00"
          c3__service-0: 0qaazv8dow-c3-c3-k8scass-cs-0010
          c3__subseq-0: "00"
          c3__updated-0: 02023-12-13T00_3A42_3A42.127Z0
          c3__updated_by-0: 0worker0
          c3__updated_from-0: 0qaazv8dow-c3-c3-k8spo-appleader-001-6648dbd5cbsq4hl0
      service: {}
    secretsProvider: internal
    tolerations:
    - effect: NoSchedule
      key: cassandra
      operator: Equal
      value: "true"
  secretsProvider: internal
status:
  conditions:
  - lastTransitionTime: "2023-08-15T18:58:42Z"
    status: "True"
    type: CassandraInitialized
  datacenters:
    cs-1855ea8cc2:
      cassandra:
        cassandraOperatorProgress: Ready
        conditions:
        - lastTransitionTime: "2023-12-13T07:23:09Z"
          message: ""
          reason: ""
          status: "True"
          type: Healthy
        - lastTransitionTime: "2023-08-15T18:58:40Z"
          message: ""
          reason: ""
          status: "False"
          type: Stopped
        - lastTransitionTime: "2023-08-15T18:58:40Z"
          message: ""
          reason: ""
          status: "False"
          type: ReplacingNodes
        - lastTransitionTime: "2024-11-22T07:12:08Z"
          message: ""
          reason: ""
          status: "False"
          type: Updating
        - lastTransitionTime: "2023-08-15T18:58:40Z"
          message: ""
          reason: ""
          status: "False"
          type: RollingRestart
        - lastTransitionTime: "2023-08-15T18:58:40Z"
          message: ""
          reason: ""
          status: "False"
          type: Resuming
        - lastTransitionTime: "2023-08-15T18:58:40Z"
          message: ""
          reason: ""
          status: "False"
          type: ScalingDown
        - lastTransitionTime: "2023-08-15T18:58:40Z"
          message: ""
          reason: ""
          status: "True"
          type: Valid
        - lastTransitionTime: "2023-08-15T18:58:40Z"
          message: ""
          reason: ""
          status: "True"
          type: Initialized
        - lastTransitionTime: "2023-08-15T18:58:40Z"
          message: ""
          reason: ""
          status: "True"
          type: Ready
        - lastTransitionTime: "2024-11-22T00:02:31Z"
          message: ""
          reason: ""
          status: "False"
          type: RequiresUpdate
        datacenterName: ""
        lastServerNodeStarted: "2024-11-22T07:11:38Z"
        nodeStatuses:
          cs-1855ea8cc2-cs-1855ea8cc2-default-sts-0:
            hostID: 975804b2-abad-425f-a780-ba8f66bec4be
          cs-1855ea8cc2-cs-1855ea8cc2-default-sts-1:
            hostID: 41c647a0-7057-460c-bf84-6d187175aefe
          cs-1855ea8cc2-cs-1855ea8cc2-default-sts-2:
            hostID: 475846a4-c5da-49e7-9dd5-35838abb81db
        observedGeneration: 64
        quietPeriod: "2024-11-22T07:12:14Z"
        superUserUpserted: "2024-11-22T07:12:09Z"
        usersUpserted: "2024-11-22T07:12:09Z"
      reaper:
        conditions:
        - lastTransitionTime: "2024-11-22T07:01:08Z"
          status: "True"
          type: Ready
        progress: Running
      stargate:
        availableReplicas: 1
        conditions:
        - lastTransitionTime: "2024-11-22T07:13:21Z"
          status: "True"
          type: Ready
        deploymentRefs:
        - cs-1855ea8cc2-cs-1855ea8cc2-default-stargate-deployment
        progress: Running
        readyReplicas: 1
        readyReplicasRatio: 1/1
        replicas: 1
        serviceRef: cs-1855ea8cc2-cs-1855ea8cc2-stargate-service
        updatedReplicas: 1
  error: None
```

MedusaBackupSchedule:
```
apiVersion: medusa.k8ssandra.io/v1alpha1
kind: MedusaBackupSchedule
metadata:
  creationTimestamp: "2023-12-13T05:13:26Z"
  generation: 1
  labels:
    app: c3aiops
    ops.c3.ai/parent-resource: qaazv8dow-c3cassandra-cs-1855ea8cc2
    role: c3aiops-C3Cassandra
  name: cs-1855ea8cc2-cs-1855ea8cc2
  namespace: qaazv8dow
  resourceVersion: "348632318"
  uid: 3c1a38e6-39dd-41ee-a0de-431f1cbfd2a4
spec:
  backupSpec:
    backupType: differential
    cassandraDatacenter: cs-1855ea8cc2
  cronSchedule: '@daily'
status:
  lastExecution: "2024-11-22T00:00:00Z"
  nextSchedule: "2024-11-23T00:00:00Z"

```

* K8ssandra Operator Logs:

```
Error:
15m         Warning   FailedCreate           statefulset/cs-1855ea8cc2-cs-1855ea8cc2-default-sts                               create Pod cs-1855ea8cc2-cs-1855ea8cc2-default-sts-0 in StatefulSet cs-1855ea8cc2-cs-1855ea8cc2-default-sts failed error: Internal error occurred: failed calling webhook "mpod.kb.io": failed to call webhook: Post "https://c3aiops-k8ssandra-operator-webhook-service.c3-opsadmin.svc:443/mutate-v1-pod-secrets-inject?timeout=10s": no endpoints available for service "c3aiops-k8ssandra-operator-webhook-service"
```

```
2024-11-22T04:25:56.431Z        INFO    Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference        {"controller": "medusabackupjob", "controllerGroup": "medusa.k8ssandra.io", "controllerKind": "MedusaBackupJob", "MedusaBackupJob": {"name":"cs-1855ea8cc2-cs-1855ea8cc2-1717977600","namespace":"qaazv8dow"}, "namespace": "qaazv8dow", "name": "cs-1855ea8cc2-cs-1855ea8cc2-1717977600", "reconcileID": "cec567f9-0bc6-4c46-9c55-5ee233259314"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x48 pc=0x55f16f305904]

goroutine 597 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:119 +0x1e5
panic({0x55f16f77d520?, 0x55f170644c80?})
        runtime/panic.go:770 +0x132
github.com/k8ssandra/k8ssandra-operator/controllers/medusa.(*MedusaBackupJobReconciler).createMedusaBackup(0xc000846480, {0x55f16f9d1440, 0xc0007c78f0}, 0xc00089b040, 0x0, {{0x55f16f9d58e8?, 0xc0007c7950?}, 0x55f170746ce0?})
        github.com/k8ssandra/k8ssandra-operator/controllers/medusa/medusabackupjob_controller.go:275 +0x5a4
github.com/k8ssandra/k8ssandra-operator/controllers/medusa.(*MedusaBackupJobReconciler).Reconcile(0xc000846480, {0x55f16f9d1440, 0xc0007c78f0}, {{{0xc000ca6356, 0x9}, {0xc000cc2420, 0x26}}})
        github.com/k8ssandra/k8ssandra-operator/controllers/medusa/medusabackupjob_controller.go:171 +0x569
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x55f16f9d1440?, {0x55f16f9d1440?, 0xc0007c78f0?}, {{{0xc000ca6356?, 0x55f16f6bee00?}, {0xc000cc2420?, 0x10?}}})
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000188a00, {0x55f16f9d1478, 0xc0001f3400}, {0x55f16f7dea00, 0xc000b161c0})
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323 +0x345
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000188a00, {0x55f16f9d1478, 0xc0001f3400})
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274 +0x1c9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 140
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:231 +0x50c
```

Logs once k8ssandra-operator reached a Running state after the above crash:
```
2024-11-22T06:32:06.089Z        ERROR   Failed to fetch datacenter pods {"controller": "k8ssandracluster", "controllerGroup": "k8ssandra.io", "controllerKind": "K8ssandraCluster", "K8ssandraCluster": {"name":"cs-1855ea8cc2","namespace":"qaazv8dow"}, "namespace": "qaazv8dow", "name": "cs-1855ea8cc2", "reconcileID": "f52dcd0e-f2fa-4de2-9cd8-a5cf249a4dd8", "K8ssandraCluster": "qaazv8dow/cs-1855ea8cc2", "CassandraDatacenter": "qaazv8dow/cs-1855ea8cc2", "K8SContext": "", "error": "no pods in READY state found in datacenter cs-1855ea8cc2"}
github.com/k8ssandra/k8ssandra-operator/pkg/cassandra.(*defaultManagementApiFacade).ListKeyspaces
        github.com/k8ssandra/k8ssandra-operator/pkg/cassandra/management.go:192
github.com/k8ssandra/k8ssandra-operator/pkg/cassandra.(*defaultManagementApiFacade).EnsureKeyspaceReplication
        github.com/k8ssandra/k8ssandra-operator/pkg/cassandra/management.go:291
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).updateReplicationOfSystemKeyspaces
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/schemas.go:161
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).checkSchemas
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/schemas.go:43
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).reconcileDatacenters
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/datacenters.go:207
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).reconcile
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/k8ssandracluster_controller.go:144
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).Reconcile
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/k8ssandracluster_controller.go:92
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235
2024-11-22T06:32:06.090Z        ERROR   Failed to update replication    {"controller": "k8ssandracluster", "controllerGroup": "k8ssandra.io", "controllerKind": "K8ssandraCluster", "K8ssandraCluster": {"name":"cs-1855ea8cc2","namespace":"qaazv8dow"}, "namespace": "qaazv8dow", "name": "cs-1855ea8cc2", "reconcileID": "f52dcd0e-f2fa-4de2-9cd8-a5cf249a4dd8", "K8ssandraCluster": "qaazv8dow/cs-1855ea8cc2", "CassandraDatacenter": "qaazv8dow/cs-1855ea8cc2", "K8SContext": "", "keyspace": "system_traces", "error": "no pods in READY state found in datacenter cs-1855ea8cc2"}
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).updateReplicationOfSystemKeyspaces
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/schemas.go:165
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).checkSchemas
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/schemas.go:43
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).reconcileDatacenters
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/datacenters.go:207
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).reconcile
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/k8ssandracluster_controller.go:144
github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra.(*K8ssandraClusterReconciler).Reconcile
        github.com/k8ssandra/k8ssandra-operator/controllers/k8ssandra/k8ssandracluster_controller.go:92
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:122
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:323
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235
2024-11-22T06:32:06.090Z        DEBUG   events  no pods in READY state found in datacenter cs-1855ea8cc2        {"type": "Warning", "object": {"kind":"K8ssandraCluster","namespace":"qaazv8dow","name":"cs-1855ea8cc2","uid":"6ffc08c2-d358-4f3b-a102-e99dbd396651","apiVersion":"k8ssandra.io/v1alpha1","resourceVersion":"348839464"}, "reason": "Reconcile Error"}
2024-11-22T06:32:06.106Z        INFO    updated k8ssandracluster status {"controller": "k8ssandracluster", "controllerGroup": "k8ssandra.io", "controllerKind": "K8ssandraCluster", "K8ssandraCluster": {"name":"cs-1855ea8cc2","namespace":"qaazv8dow"}, "namespace": "qaazv8dow", "name": "cs-1855ea8cc2", "reconcileID": "f52dcd0e-f2fa-4de2-9cd8-a5cf249a4dd8", "K8ssandraCluster": "qaazv8dow/cs-1855ea8cc2"}
2024-11-22T06:32:06.106Z        ERROR   Reconciler error        {"controller": "k8ssandracluster", "controllerGroup": "k8ssandra.io", "controllerKind": "K8ssandraCluster", "K8ssandraCluster": {"name":"cs-1855ea8cc2","namespace":"qaazv8dow"}, "namespace": "qaazv8dow", "name": "cs-1855ea8cc2", "reconcileID": "f52dcd0e-f2fa-4de2-9cd8-a5cf249a4dd8", "error": "no pods in READY state found in datacenter cs-1855ea8cc2"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:329
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:274
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
        sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:235
```

**Anything else we need to know?**:
I checked the code and I believe all k8ssandra-operator versions are impacted, not only 1.14

┆Issue is synchronized with this [Jira Story](https://datastax.jira.com/browse/K8OP-294) by [Unito](https://www.unito.io)
┆Issue Number: K8OP-294

--- a/CHANGELOG/CHANGELOG-1.21.md
+++ b/CHANGELOG/CHANGELOG-1.21.md
@@ -19,3 +19,4 @@ When cutting a new release, update the `unreleased` heading to the tag being gen
 * [BUGFIX] [#1383](https://github.com/k8ssandra/k8ssandra-operator/issues/1383) Do not create MedusaBackup if MedusaBakupJob did not fully succeed
 * [ENHANCEMENT] [#1667](https://github.com/k8ssahttps://github.com/k8ssandra/k8ssandra/issues/1667) Add `skipSchemaMigration` option to `K8ssandraCluster.spec.reaper`
 * [FEATURE] [#1034](https://github.com/k8ssandra/k8ssandra-operator/issues/1034) Add support for priorityClassName
+* [BUGFIX] [#1454](https://github.com/k8ssandra/k8ssandra-operator/issues/1454) Do not try to work out backup status if there are no pods

--- a/controllers/medusa/medusabackupjob_controller.go
+++ b/controllers/medusa/medusabackupjob_controller.go
@@ -18,7 +18,6 @@ package medusa
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"net"
 	"strings"
@@ -34,6 +33,7 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	"github.com/go-logr/logr"
 	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
@@ -253,7 +253,7 @@ func (r *MedusaBackupJobReconciler) getBackupSummary(ctx context.Context, backup
 			}
 		}
 	}
-	return nil, errors.New("backup summary couldn't be found")
+	return nil, reconcile.TerminalError(fmt.Errorf("backup summary couldn't be found"))
 }
 
 func (r *MedusaBackupJobReconciler) createMedusaBackup(ctx context.Context, backup *medusav1alpha1.MedusaBackupJob, backupSummary *medusa.BackupSummary, logger logr.Logger) error {

--- a/controllers/medusa/medusabackupjob_controller.go
+++ b/controllers/medusa/medusabackupjob_controller.go
@@ -103,6 +103,10 @@ func (r *MedusaBackupJobReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		logger.Error(err, "Failed to get datacenter pods")
 		return ctrl.Result{}, err
 	}
+	if len(pods) == 0 {
+		logger.Info("No pods found for datacenter", "CassandraDatacenter", cassdcKey)
+		return ctrl.Result{Requeue: true}, nil
+	}
 
 	// If there is anything in progress, simply requeue the request until each pod has finished or errored
 	if len(backupJob.Status.InProgress) > 0 {
@@ -185,6 +189,11 @@ func (r *MedusaBackupJobReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		if err != nil {
 			logger.Error(err, "Failed to get backup summary")
 			return ctrl.Result{}, err
+		}
+		if backupSummary == nil {
+			// if a backup is complete, but we fail to find the summary, we're in a non-recoverable situation
+			logger.Info("Backup summary not found", "backupJob", backupJob)
+			return ctrl.Result{Requeue: false}, nil
 		}
 		if err := r.createMedusaBackup(ctx, backupJob, backupSummary, logger); err != nil {
 			logger.Error(err, "Failed to create MedusaBackup")

--- a/controllers/medusa/medusabackupjob_controller.go
+++ b/controllers/medusa/medusabackupjob_controller.go
@@ -245,6 +245,11 @@ func (r *MedusaBackupJobReconciler) getBackupSummary(ctx context.Context, backup
 			return nil, err
 		} else {
 			for _, remoteBackup := range remoteBackups {
+				if remoteBackup == nil {
+					err := fmt.Errorf("backup %s summary is nil", backup.Name)
+					logger.Error(err, "remote backup is nil")
+					return nil, err
+				}
 				logger.Info("found backup", "CassandraPod", pod.Name, "Backup", remoteBackup.BackupName)
 				if backup.ObjectMeta.Name == remoteBackup.BackupName {
 					return remoteBackup, nil

--- a/controllers/medusa/medusabackupjob_controller.go
+++ b/controllers/medusa/medusabackupjob_controller.go
@@ -18,12 +18,13 @@ package medusa
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"net"
 	"strings"
 
 	corev1 "k8s.io/api/core/v1"
-	"k8s.io/apimachinery/pkg/api/errors"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
@@ -67,7 +68,7 @@ func (r *MedusaBackupJobReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	err := r.Get(ctx, req.NamespacedName, instance)
 	if err != nil {
 		logger.Error(err, "Failed to get MedusaBackupJob")
-		if errors.IsNotFound(err) {
+		if apiErrors.IsNotFound(err) {
 			return ctrl.Result{}, nil
 		}
 		return ctrl.Result{}, err
@@ -190,11 +191,6 @@ func (r *MedusaBackupJobReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 			logger.Error(err, "Failed to get backup summary")
 			return ctrl.Result{}, err
 		}
-		if backupSummary == nil {
-			// if a backup is complete, but we fail to find the summary, we're in a non-recoverable situation
-			logger.Info("Backup summary not found", "backupJob", backupJob)
-			return ctrl.Result{Requeue: false}, nil
-		}
 		if err := r.createMedusaBackup(ctx, backupJob, backupSummary, logger); err != nil {
 			logger.Error(err, "Failed to create MedusaBackup")
 			return ctrl.Result{}, err
@@ -257,7 +253,7 @@ func (r *MedusaBackupJobReconciler) getBackupSummary(ctx context.Context, backup
 			}
 		}
 	}
-	return nil, nil
+	return nil, errors.New("backup summary couldn't be found")
 }
 
 func (r *MedusaBackupJobReconciler) createMedusaBackup(ctx context.Context, backup *medusav1alpha1.MedusaBackupJob, backupSummary *medusa.BackupSummary, logger logr.Logger) error {
@@ -266,7 +262,7 @@ func (r *MedusaBackupJobReconciler) createMedusaBackup(ctx context.Context, back
 	backupKey := types.NamespacedName{Namespace: backup.ObjectMeta.Namespace, Name: backup.Name}
 	backupResource := &medusav1alpha1.MedusaBackup{}
 	if err := r.Get(ctx, backupKey, backupResource); err != nil {
-		if errors.IsNotFound(err) {
+		if apiErrors.IsNotFound(err) {
 			// Backup doesn't exist, create it
 			startTime := backup.Status.StartTime
 			finishTime := metav1.Now()
@@ -340,7 +336,7 @@ func backupStatus(ctx context.Context, name string, pod *corev1.Pod, clientFacto
 		resp, err := medusaClient.BackupStatus(ctx, name)
 		if err != nil {
 			// the gRPC client does not return proper NotFound error, we need to check the payload too
-			if errors.IsNotFound(err) || strings.Contains(err.Error(), "NotFound") {
+			if apiErrors.IsNotFound(err) || strings.Contains(err.Error(), "NotFound") {
 				logger.Info(fmt.Sprintf("did not find backup %s for pod %s", name, pod.Name))
 				return medusa.StatusType_UNKNOWN, nil
 			}

--- a/controllers/medusa/medusabackupjob_controller_test.go
+++ b/controllers/medusa/medusabackupjob_controller_test.go
@@ -433,35 +433,40 @@ func (c *fakeMedusaClient) GetBackups(ctx context.Context) ([]*medusa.BackupSumm
 			status = *medusa.StatusType_IN_PROGRESS.Enum()
 		}
 
-		backup := &medusa.BackupSummary{
-			BackupName:    name,
-			StartTime:     0,
-			FinishTime:    10,
-			TotalNodes:    3,
-			FinishedNodes: 3,
-			TotalObjects:  fakeBackupFileCount,
-			TotalSize:     fakeBackupByteSize,
-			Status:        status,
-			Nodes: []*medusa.BackupNode{
-				{
-					Host:       "host1",
-					Tokens:     []int64{1, 2, 3},
-					Datacenter: "dc1",
-					Rack:       "rack1",
+		var backup *medusa.BackupSummary
+		if strings.HasPrefix(name, backupWithNoPods) {
+			backup = nil
+		} else {
+			backup = &medusa.BackupSummary{
+				BackupName:    name,
+				StartTime:     0,
+				FinishTime:    10,
+				TotalNodes:    3,
+				FinishedNodes: 3,
+				TotalObjects:  fakeBackupFileCount,
+				TotalSize:     fakeBackupByteSize,
+				Status:        status,
+				Nodes: []*medusa.BackupNode{
+					{
+						Host:       "host1",
+						Tokens:     []int64{1, 2, 3},
+						Datacenter: "dc1",
+						Rack:       "rack1",
+					},
+					{
+						Host:       "host2",
+						Tokens:     []int64{1, 2, 3},
+						Datacenter: "dc1",
+						Rack:       "rack1",
+					},
+					{
+						Host:       "host3",
+						Tokens:     []int64{1, 2, 3},
+						Datacenter: "dc1",
+						Rack:       "rack1",
+					},
 				},
-				{
-					Host:       "host2",
-					Tokens:     []int64{1, 2, 3},
-					Datacenter: "dc1",
-					Rack:       "rack1",
-				},
-				{
-					Host:       "host3",
-					Tokens:     []int64{1, 2, 3},
-					Datacenter: "dc1",
-					Rack:       "rack1",
-				},
-			},
+			}
 		}
 		backups = append(backups, backup)
 	}

--- a/controllers/medusa/medusabackupjob_controller_test.go
+++ b/controllers/medusa/medusabackupjob_controller_test.go
@@ -7,6 +7,7 @@ import (
 	"strings"
 	"sync"
 	"testing"
+	"time"
 
 	cassdcapi "github.com/k8ssandra/cass-operator/apis/cassandra/v1beta1"
 	k8ss "github.com/k8ssandra/k8ssandra-operator/apis/k8ssandra/v1alpha1"
@@ -31,6 +32,7 @@ const (
 	successfulBackupName = "good-backup"
 	failingBackupName    = "bad-backup"
 	missingBackupName    = "missing-backup"
+	backupWithNoPods     = "backup-with-no-pods"
 	dc1PodPrefix         = "192.168.1."
 	dc2PodPrefix         = "192.168.2."
 	fakeBackupFileCount  = int64(13)
@@ -167,6 +169,10 @@ func testMedusaBackupDatacenter(t *testing.T, ctx context.Context, f *framework.
 	backupCreated = createAndVerifyMedusaBackup(dc1Key, dc1, f, ctx, require, t, namespace, missingBackupName)
 	require.False(backupCreated, "the backup object shouldn't have been created")
 
+	// in K8OP-294 we found out we can try to make backups on StatefulSets with no pods
+	backupCreated = createAndVerifyMedusaBackup(dc1Key, dc1, f, ctx, require, t, namespace, backupWithNoPods)
+	require.False(backupCreated, "the backup object shouldn't have been created")
+
 	err = f.DeleteK8ssandraCluster(ctx, client.ObjectKey{Namespace: kc.Namespace, Name: kc.Name}, timeout, interval)
 	require.NoError(err, "failed to delete K8ssandraCluster")
 	verifyObjectDoesNotExist(ctx, t, f, dc1Key, &cassdcapi.CassandraDatacenter{})
@@ -202,13 +208,23 @@ func createAndVerifyMedusaBackup(dcKey framework.ClusterKey, dc *cassdcapi.Cassa
 		}
 	}
 
-	createDatacenterPods(t, f, ctx, dcKey, dc)
+	if backupName != backupWithNoPods {
+		createDatacenterPods(t, f, ctx, dcKey, dc)
+	}
 
 	dcCopy := dc.DeepCopy()
 	dcKeyCopy := framework.NewClusterKey(f.DataPlaneContexts[0], dcKey.Namespace+"-copy", dcKey.Name)
 	dcCopy.ObjectMeta.Namespace = dc.Namespace + "-copy"
 
-	createDatacenterPods(t, f, ctx, dcKeyCopy, dcCopy)
+	if backupName != backupWithNoPods {
+		createDatacenterPods(t, f, ctx, dcKeyCopy, dcCopy)
+	}
+
+	// clean up the pods to keep the environment reusable
+	if backupName != backupWithNoPods {
+		defer deleteDatacenterPods(t, f, ctx, dcKey, dc)
+		defer deleteDatacenterPods(t, f, ctx, dcKeyCopy, dc)
+	}
 
 	t.Log("creating MedusaBackupJob")
 	backupKey := framework.NewClusterKey(dcKey.K8sContext, dcKey.Namespace, backupName)
@@ -225,32 +241,14 @@ func createAndVerifyMedusaBackup(dcKey framework.ClusterKey, dc *cassdcapi.Cassa
 	err := f.Create(ctx, backupKey, backup)
 	require.NoError(err, "failed to create MedusaBackupJob")
 
-	t.Log("verify that the backups are started")
-	require.Eventually(func() bool {
-		t.Logf("Requested backups: %v", medusaClientFactory.GetRequestedBackups(dc.DatacenterName()))
-		updated := &api.MedusaBackupJob{}
-		err := f.Get(ctx, backupKey, updated)
-		if err != nil {
-			t.Logf("failed to get MedusaBackupJob: %v", err)
-			return false
-		}
-		return !updated.Status.StartTime.IsZero()
-	}, timeout, interval)
-
-	t.Log("verify the backup finished")
-	require.Eventually(func() bool {
-		t.Logf("Requested backups: %v", medusaClientFactory.GetRequestedBackups(dc.DatacenterName()))
-		updated := &api.MedusaBackupJob{}
-		err := f.Get(ctx, backupKey, updated)
-		if err != nil {
-			return false
-		}
-		t.Logf("backup %s finish time: %v", backupName, updated.Status.FinishTime)
-		t.Logf("backup %s failed: %v", backupName, updated.Status.Failed)
-		t.Logf("backup %s finished: %v", backupName, updated.Status.Finished)
-		t.Logf("backup %s in progress: %v", backupName, updated.Status.InProgress)
-		return !updated.Status.FinishTime.IsZero()
-	}, timeout, interval)
+	if backupName != backupWithNoPods {
+		t.Log("verify that the backups start eventually")
+		verifyBackupJobStarted(require.Eventually, t, dc, f, ctx, backupKey)
+		verifyBackupJobFinished(t, require, dc, f, ctx, backupKey, backupName)
+	} else {
+		t.Log("verify that the backups never start")
+		verifyBackupJobStarted(require.Never, t, dc, f, ctx, backupKey)
+	}
 
 	t.Log("check for the MedusaBackup being created")
 	medusaBackupKey := framework.NewClusterKey(dcKey.K8sContext, dcKey.Namespace, backupName)
@@ -258,6 +256,7 @@ func createAndVerifyMedusaBackup(dcKey framework.ClusterKey, dc *cassdcapi.Cassa
 	err = f.Get(ctx, medusaBackupKey, medusaBackup)
 	if err != nil {
 		if errors.IsNotFound(err) {
+			// exit the test if the backup was not created. this happens for some backup names on purpose
 			return false
 		}
 	}
@@ -272,6 +271,43 @@ func createAndVerifyMedusaBackup(dcKey framework.ClusterKey, dc *cassdcapi.Cassa
 	require.Equal(int(dc.Spec.Size), len(medusaClientFactory.GetRequestedBackups(dc.DatacenterName())))
 
 	return true
+}
+
+func verifyBackupJobFinished(t *testing.T, require *require.Assertions, dc *cassdcapi.CassandraDatacenter, f *framework.Framework, ctx context.Context, backupKey framework.ClusterKey, backupName string) {
+	t.Log("verify the backup finished")
+	require.Eventually(func() bool {
+		t.Logf("Requested backups: %v", medusaClientFactory.GetRequestedBackups(dc.DatacenterName()))
+		updated := &api.MedusaBackupJob{}
+		err := f.Get(ctx, backupKey, updated)
+		if err != nil {
+			return false
+		}
+		t.Logf("backup %s finish time: %v", backupName, updated.Status.FinishTime)
+		t.Logf("backup %s failed: %v", backupName, updated.Status.Failed)
+		t.Logf("backup %s finished: %v", backupName, updated.Status.Finished)
+		t.Logf("backup %s in progress: %v", backupName, updated.Status.InProgress)
+		return !updated.Status.FinishTime.IsZero()
+	}, timeout, interval)
+}
+
+func verifyBackupJobStarted(
+	verifyFunction func(condition func() bool, waitFor time.Duration, tick time.Duration, msgAndArgs ...interface{}),
+	t *testing.T,
+	dc *cassdcapi.CassandraDatacenter,
+	f *framework.Framework,
+	ctx context.Context,
+	backupKey framework.ClusterKey,
+) {
+	verifyFunction(func() bool {
+		t.Logf("Requested backups: %v", medusaClientFactory.GetRequestedBackups(dc.DatacenterName()))
+		updated := &api.MedusaBackupJob{}
+		err := f.Get(ctx, backupKey, updated)
+		if err != nil {
+			t.Logf("failed to get MedusaBackupJob: %v", err)
+			return false
+		}
+		return !updated.Status.StartTime.IsZero()
+	}, timeout, interval)
 }
 
 func reconcileReplicatedSecret(ctx context.Context, t *testing.T, f *framework.Framework, kc *k8ss.K8ssandraCluster) {
@@ -486,6 +522,26 @@ func findDatacenterCondition(status *cassdcapi.CassandraDatacenterStatus, condTy
 		}
 	}
 	return nil
+}
+
+func deleteDatacenterPods(t *testing.T, f *framework.Framework, ctx context.Context, dcKey framework.ClusterKey, dc *cassdcapi.CassandraDatacenter) {
+	for i := int32(0); i < dc.Spec.Size; i++ {
+		pod := &corev1.Pod{}
+		podName := fmt.Sprintf("%s-%s-%d", dc.Spec.ClusterName, dc.DatacenterName(), i)
+		podKey := framework.NewClusterKey(dcKey.K8sContext, dcKey.Namespace, podName)
+		err := f.Get(ctx, podKey, pod)
+		if err != nil {
+			if !errors.IsNotFound(err) {
+				t.Logf("failed to get pod %s: %v", podKey, err)
+			}
+		} else {
+			err = f.Delete(ctx, podKey, pod)
+			if err != nil {
+				t.Logf("failed to delete pod %s: %v", podKey, err)
+			}
+		}
+
+	}
 }
 
 func createDatacenterPods(t *testing.T, f *framework.Framework, ctx context.Context, dcKey framework.ClusterKey, dc *cassdcapi.CassandraDatacenter) {

--- a/controllers/medusa/medusabackupjob_controller_test.go
+++ b/controllers/medusa/medusabackupjob_controller_test.go
@@ -526,7 +526,7 @@ func findDatacenterCondition(status *cassdcapi.CassandraDatacenterStatus, condTy
 }
 
 func deleteDatacenterPods(t *testing.T, f *framework.Framework, ctx context.Context, dcKey framework.ClusterKey, dc *cassdcapi.CassandraDatacenter) {
-	for i := int32(0); i < dc.Spec.Size; i++ {
+	for i := 0; i < int(dc.Spec.Size); i++ {
 		pod := &corev1.Pod{}
 		podName := fmt.Sprintf("%s-%s-%d", dc.Spec.ClusterName, dc.DatacenterName(), i)
 		podKey := framework.NewClusterKey(dcKey.K8sContext, dcKey.Namespace, podName)

--- a/controllers/medusa/medusabackupjob_controller_test.go
+++ b/controllers/medusa/medusabackupjob_controller_test.go
@@ -535,18 +535,10 @@ func deleteDatacenterPods(t *testing.T, f *framework.Framework, ctx context.Cont
 		pod := &corev1.Pod{}
 		podName := fmt.Sprintf("%s-%s-%d", dc.Spec.ClusterName, dc.DatacenterName(), i)
 		podKey := framework.NewClusterKey(dcKey.K8sContext, dcKey.Namespace, podName)
-		err := f.Get(ctx, podKey, pod)
+		err := f.Delete(ctx, podKey, pod)
 		if err != nil {
-			if !errors.IsNotFound(err) {
-				t.Logf("failed to get pod %s: %v", podKey, err)
-			}
-		} else {
-			err = f.Delete(ctx, podKey, pod)
-			if err != nil {
-				t.Logf("failed to delete pod %s: %v", podKey, err)
-			}
+			t.Logf("failed to delete pod %s: %v", podKey, err)
 		}
-
 	}
 }
 

--- a/controllers/medusa/medusabackupjob_controller_test.go
+++ b/controllers/medusa/medusabackupjob_controller_test.go
@@ -249,7 +249,10 @@ func createAndVerifyMedusaBackup(dcKey framework.ClusterKey, dc *cassdcapi.Cassa
 	if backupName != backupWithNoPods {
 		t.Log("verify that the backups start eventually")
 		verifyBackupJobStarted(require.Eventually, t, dc, f, ctx, backupKey)
-		verifyBackupJobFinished(t, require, dc, f, ctx, backupKey, backupName)
+		if backupName != backupWithNilSummary {
+			// backup that will receive a nil summary will never finish, can't assert on that
+			verifyBackupJobFinished(t, require, dc, f, ctx, backupKey, backupName)
+		}
 	} else {
 		t.Log("verify that the backups never start")
 		verifyBackupJobStarted(require.Never, t, dc, f, ctx, backupKey)
@@ -536,9 +539,12 @@ func findDatacenterCondition(status *cassdcapi.CassandraDatacenterStatus, condTy
 
 func deleteDatacenterPods(t *testing.T, f *framework.Framework, ctx context.Context, dcKey framework.ClusterKey, dc *cassdcapi.CassandraDatacenter) {
 	for i := 0; i < int(dc.Spec.Size); i++ {
-		pod := &corev1.Pod{}
+
 		podName := fmt.Sprintf("%s-%s-%d", dc.Spec.ClusterName, dc.DatacenterName(), i)
 		podKey := framework.NewClusterKey(dcKey.K8sContext, dcKey.Namespace, podName)
+		pod := &corev1.Pod{
+			ObjectMeta: metav1.ObjectMeta{Name: podKey.Name, Namespace: podKey.Namespace},
+		}
 		err := f.Delete(ctx, podKey, pod)
 		if err != nil {
 			t.Logf("failed to delete pod %s: %v", podKey, err)

--- a/controllers/medusa/medusabackupjob_controller_test.go
+++ b/controllers/medusa/medusabackupjob_controller_test.go
@@ -208,22 +208,23 @@ func createAndVerifyMedusaBackup(dcKey framework.ClusterKey, dc *cassdcapi.Cassa
 		}
 	}
 
-	if backupName != backupWithNoPods {
-		createDatacenterPods(t, f, ctx, dcKey, dc)
-	}
+	createDatacenterPods(t, f, ctx, dcKey, dc)
 
 	dcCopy := dc.DeepCopy()
 	dcKeyCopy := framework.NewClusterKey(f.DataPlaneContexts[0], dcKey.Namespace+"-copy", dcKey.Name)
 	dcCopy.ObjectMeta.Namespace = dc.Namespace + "-copy"
 
-	if backupName != backupWithNoPods {
-		createDatacenterPods(t, f, ctx, dcKeyCopy, dcCopy)
-	}
+	createDatacenterPods(t, f, ctx, dcKeyCopy, dcCopy)
 
-	// clean up the pods to keep the environment reusable
-	if backupName != backupWithNoPods {
-		defer deleteDatacenterPods(t, f, ctx, dcKey, dc)
-		defer deleteDatacenterPods(t, f, ctx, dcKeyCopy, dc)
+	// one test scenario needs to have no pods available in the STSs (see #1454)
+	// we reproduce that by deleting the pods. we need this for the medusa backup controller tests to work
+	// however, we need to bring them back up because medusa task controller tests interact with them later
+	// both backup and task controller tests use this function to verify backups
+	if backupName == backupWithNoPods {
+		deleteDatacenterPods(t, f, ctx, dcKey, dc)
+		deleteDatacenterPods(t, f, ctx, dcKeyCopy, dc)
+		defer createDatacenterPods(t, f, ctx, dcKey, dc)
+		defer createDatacenterPods(t, f, ctx, dcKeyCopy, dcCopy)
 	}
 
 	t.Log("creating MedusaBackupJob")

--- a/controllers/medusa/medusabackupjob_controller_test.go
+++ b/controllers/medusa/medusabackupjob_controller_test.go
@@ -33,6 +33,7 @@ const (
 	failingBackupName    = "bad-backup"
 	missingBackupName    = "missing-backup"
 	backupWithNoPods     = "backup-with-no-pods"
+	backupWithNilSummary = "backup-with-nil-summary"
 	dc1PodPrefix         = "192.168.1."
 	dc2PodPrefix         = "192.168.2."
 	fakeBackupFileCount  = int64(13)
@@ -171,6 +172,9 @@ func testMedusaBackupDatacenter(t *testing.T, ctx context.Context, f *framework.
 
 	// in K8OP-294 we found out we can try to make backups on StatefulSets with no pods
 	backupCreated = createAndVerifyMedusaBackup(dc1Key, dc1, f, ctx, require, t, namespace, backupWithNoPods)
+	require.False(backupCreated, "the backup object shouldn't have been created")
+	// in that same effort, we also found out we can have nil instead of backup sumamry
+	backupCreated = createAndVerifyMedusaBackup(dc1Key, dc1, f, ctx, require, t, namespace, backupWithNilSummary)
 	require.False(backupCreated, "the backup object shouldn't have been created")
 
 	err = f.DeleteK8ssandraCluster(ctx, client.ObjectKey{Namespace: kc.Namespace, Name: kc.Name}, timeout, interval)
@@ -434,7 +438,7 @@ func (c *fakeMedusaClient) GetBackups(ctx context.Context) ([]*medusa.BackupSumm
 		}
 
 		var backup *medusa.BackupSummary
-		if strings.HasPrefix(name, backupWithNoPods) {
+		if strings.HasPrefix(name, backupWithNilSummary) {
 			backup = nil
 		} else {
 			backup = &medusa.BackupSummary{

--- a/pkg/medusa/utils.go
+++ b/pkg/medusa/utils.go
@@ -2,6 +2,7 @@ package medusa
 
 import (
 	"context"
+	"errors"
 
 	"github.com/go-logr/logr"
 	corev1 "k8s.io/api/core/v1"
@@ -21,8 +22,8 @@ func GetCassandraDatacenterPods(ctx context.Context, cassdc *cassdcapi.Cassandra
 		return nil, err
 	}
 
-	pods := make([]corev1.Pod, 0)
-	pods = append(pods, podList.Items...)
-
-	return pods, nil
+	if podList.Items != nil {
+		return podList.Items, nil
+	}
+	return nil, errors.New("podList came with nil Items field")
 }

